### ISSUE TITLE
[SDL 0058] Add video streaming capabilities

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_ios#628]
Proposal: [Add video streaming capabilities](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0058-video-streaming-capabilities.md)

> This proposal extends SDL video streaming feature by 1) notifying video streaming capabilities of HMI to SDL proxy, and 2) adding "video format negotiation" procedure.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/176

Steering Committee Decision:

> The Steering Committee has decided to accept this proposal with revisions. The Xevo team (@shoamano83 @jhludwig) revised, based on the discussion in the Steering Committee meeting:
1. Codecs and streaming transports need information in their descriptions to let develops know which ones are currently supported and which ones aren’t.
2. Xevo will create first draft of the fallback streaming combination order.

> The following changes were incorporated into the proposal:
1. Added a note that not all video formats are supported yet.
2. Added a suggested mechanism to pick a video format during video negotiation, including a fall back flow.

**Comment by joeljfischer** [comment]
This is blocked, waiting on a core contribution for testing

**Comment by shoamano83** [comment]
Hi @joeljfischer, @brettywhite thank you for working on this.
I plan to contribute implementations which are included in my video streaming capabilities proposal but not yet in the proxy.
Could you kindly tell me whether I should create an issue ticket for my work since this seems already closed?

FYI, the implementation will include:
- receiving GetSystemCapability response in SDLStreamingMediaLifecycleManager
- adding parameters (resolution, codec and format) when sending StartService for video streaming
- adding retry sequence when proxy receives StartService NACK

Since these are for 5.0 features, I'm currently working based on feature/streaming_media_manager branch.

**Comment by joeljfischer** [comment]
Hi @shoamano83, I would recommend holding off on this for now. Because the feature/SMM branch is not yet merged, there may be additional changes that will break what you're doing. In addition, I'm still working on parts 1 & 2 for v4.7 and would recommend that you hold off to see how I'm doing it there. I'm not sure #3 is what we want to do, and if it is, it will require a fair bit of thought on how fallbacks should work. Finally, this is a fairly complex high-level feature, so I'm going to be fairly exacting (read: anal) about how this is implemented. It will be difficult to change this behavior in the future, so be aware that if you take this on, you should either: 1. Run a design spec through me first, or 2. Have time set aside to redo aspects of it if I object to the design choices.

**Comment by joeljfischer** [comment]
@shoamano83 I'm doing the v4.7 updates as part of the constructed payloads ticket. The v5.0 work will probably be on the same ticket. If you're looking to do more, such as your third bullet, that would have to be a part of a separate ticket (and proposal).

**Comment by joeljfischer** [comment]
@shoamano83 After considering it more, I would prefer to tackle this myself. Your third point will not be included, but 1 & 2 will be. If you'd like to propose something for 3, it will have to go through the proposal process anyway.

**Comment by shoamano83** [comment]
Hi @joeljfischer, OK thank you for your work on it!

Regarding #3, I meant following section in the proposal. Sorry for your confusion.

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0058-video-streaming-capabilities.md
> When selecting a video format, SDL proxy should pick a format that is most preferable to the head unit and is also supported by the proxy itself. If HMI should reply a negative response with the format, then the proxy picks the second preferable format and try again. SDL proxy can continue this routine until it gets to the end of video format list. If none of the format is accepted (which is unlikely to happen), then the proxy may emit Start Service Control Frame without any video configuration parameters, to fall back to existing sequence.

**Comment by joeljfischer** [comment]
@shoamano83 Oh, okay. Though I have to admit, I'm rather confused as to why the head unit would NAK after saying that it supports a configuration of protocol and codec.

**Comment by shoamano83** [comment]
Honestly speaking, I don't think HMI will throw an error when we just switch between RAW and RTP formats. (As you know, I was requested to add this mechanism during SDLC review.)

The mechanism *might* be useful in future when SDL supports multiple video codecs. For example, a head unit has only one hardware H.265 decoder, and it is occupied with another usage (like native media player playing H.265 video). In such case, HMI may want to temporarily reject H.265 stream but accept H.264 instead. Again, I'm not sure if we will have such use-case.
